### ISSUE TITLE
DataObject.GetData(FileContents) returns null

### ISSUE THREAD

**Issue by asklar** [dotnet/winforms#3026]
* .NET Core Version: 
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  3.1.201 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**
IDataObject.GetFormats returns FileContents, but IDataObject.GetData("FileContents") returns null

**Expected behavior:**
Returns the file contents

**Minimal repro:**

**Comment by weltkante** [comment]
I never used this particular format myself so I can't guess which application you might have used. You'll probably have to be more specific in how to repro, since DataObject content is highly dependent on how the source application presents it data.

**Comment by asklar** [comment]
> I never used this particular format myself so I can't guess which application you might have used. You'll probably have to be more specific in how to repro, since DataObject content is highly dependent on how the source application presents it data.

Select any file in File Explorer, ctrl+C
The available formats are:
Shell IDList Array, DataObjectAttributes, DataObjectAttributesRequiringElevation, Shell Object Offsets, Preferred DropEffect, AsyncFlag, FileDrop, FileNameW, FileName, FileContents, FileGroupDescriptorW

However, the FileContents format returns null.

**Comment by weltkante** [comment]
Thanks, I can repro the null return value, but it happens both in Desktop Framework and .NET Core.

From debugging the .NET Core implementation I see that the clipboard returns a `0x800401D3 (CLIPBRD_E_BAD_DATA)` exception when reading the file data. From what I can tell this is either a bug in Windows or the source application (which would be explorer) and I doubt there's anything WinForms can do if the clipboard source doesn't provide data for a format (even if it announced it had data it can still fail to deliver it, or deliver it in a format WinForms doesn't understand)

I tried various different files in case this was file-dependent but I could not get it to work, neither in Desktop Framework nor in .NET Core. Do you have a scenario which works for you in Desktop Framework? because you said 
> Have you experienced this same bug with .NET Framework?: No

**Comment by asklar** [comment]
bump to keep this from being closed. what's the latest on this?

**Comment by RussKie** [comment]
Nice! Do you have a fix suggestion?

**Comment by asklar** [comment]
> Nice! Do you have a fix suggestion?

@RussKie Yeah I have a full implementation of clipboard APIs - including metafiles, EMF and file streams, here:
https://github.com/asklar/clipview

Specifically see https://github.com/asklar/clipview/blob/master/src/Clipboard/DataObject.cs the parts about `FileContents` and `FileGroupDescriptor` (`DataObject.InternalGetData`).

**Comment by JhonnyPala** [comment]
Hello to everyone, I have a question.
I need to copy one or more files from RDP using clipboard. Pressing Ctrl+C on remote PC affects the local clipboard and vice-versa: `FileGroupDescriptorW` is populated correctly but `FileContents `contains only the content of the first file.

Analyzing the memory with HxD, after having called `GetDataObject()`, I see that there was no trace of the other file's contents adjoining the first.

I'm talking about 3 files: PIANURA.txt, COLLINA.txt, MONTAGNA.txt; the content is the same as the name, minus the _.txt_ extension.
This is what I got by loading all the data (UTF-8 encoding), according to the GetFormats function:
```
[49344]FileGroupDescriptorW: [lenght: 1780] -> d@ ܑ/U���PIANURA.txtd@ ���P���COLLINA.txtd@ ��N���MONTAGNA.txt
[49279]FileContents: [lenght: 7] -> PIANURA
[49349]Preferred DropEffect: [lenght: 4] -> 
[49475]CanIncludeInClipboardHistory: [lenght: 4] -> 
[49474]CanUploadToCloudClipboard: [lenght: 4] -> 

```

Does anyone have experience with this stuff?
